### ISSUE TITLE
Focus appearance is impresise and fairly complicated

### ISSUE THREAD

**Issue by WilcoFiers** [w3c/wcag#1718]
In my opinion, the "area" requirement of focus appearance has gotten significantly more complicated than it can be. Specifically, I think it has the two following issues:

1. Exactly how to find the minimum areas can be ambiguous. It isn't clearly just about what the visible parts of a component are, or just the bounding box. Because the SC needs to accommodate all sorts of shapes, it got fairly complex
2. The 2 CSS pixel minimum thickness is a fairly complicated exception.

What follows is a far simpler solution that gets rid of all the little edge cases the current "area" proposal takes on. This proposal can handle all the edge cases of possibly failing things that shouldn't be failed, but makes a trade-off in that some of the more unusual edge cases that could be problematic might pass this SC.

This is essentially a different trade-off, it chooses simplicity over taking on a dozen or so extra edge cases. I think that is reasonable. Even the current proposal doesn't handle all edge cases, for example a 2px underline might pass where as a 1px double line would fail. The reason I used "3 CSS pixels" here is so that a 1px border on a pixel would pass (which has a border size of pi * diameter). 

## Suggestion

When user interface components receive keyboard focus, all the following are true for the focus indicator:
- **Indicator contrast**: ...
- **Area**: The contrasting area is at least as large as a 3 CSS pixels border along the shortest side of the unfocused component box.
- ...

### Component box
The smallest rectangle, aligned on the X and Y-axis, that can be drawn around the visible parts content of a component. Shadows, highlights and other effects around a component that largely blend into the background are not considered to be part of a component. Where a component consists of disconnected parts, such as a link that wraps onto multiple lines, each part has its own component box.

**Note**: In CSS the box of a component without border or background usually corresponds to the content box. If the element has a border or background, this is the border-box instead. If the element uses outline or box-shadow to simulate the effect of a border, the component box extends to the simulated border.

**Comment by mraccess77** [comment]
Hi @WilcoFiers I see a ... - Is there always a >= 3:1 contrast minimum with adjacent colors regardless of the width of the indicator?

**Comment by alastc** [comment]
> Area: The contrasting area is at least as large as a 3 CSS pixels border along the shortest side of the unfocused component box.

This is significantly less than the current requirement, for example, it would allow a 1px bottom border on any button which is 3x as wide as it is high, which is most text-based links/buttons.

<img src="https://user-images.githubusercontent.com/216630/113788467-e15d1700-9734-11eb-9ef8-2490a8c955b2.png" alt="Button which is 100px wide, 25px tall, with a 1px wide bottom border." width="100px">

That's making it worse for typical cases, rather than edge cases.

IMHO, it is also harder to understand than the perimeter version. It looks simpler (being 1 bullet), but then requires more calculation to work out for any particular example.

**Comment by mraccess77** [comment]
I agree with Alastair that changing the area to only 3x the shortest side without any requirement for thickness significantly reduces the benefits of this criterion.

**Comment by johnfoliot** [comment]
+1 to Jon A.

I'd expect a minimum thickness of 2px under any circumstance - but Wilco's
text MAY be able to accommodate that as well.

I am in favor of simplifying this if we can: I noted that after the
discussion Tuesday I had "fuzz-brain", and as somebody else noted, if we
experts are struggling to fully understand what we're asking for, how can
we expect non-SMEs to understand?

JF

**Comment by alastc** [comment]
> but Wilco's text MAY be able to accommodate that as well.

Perhaps, we'd need to see it. There needs to be some sort of see-saw that allows for smaller but thicker indicators, so it isn't a simple thing to do.

I think we can help with the fuzzy-brain issue by presenting several common solutions at the top of the understanding doc. E.g.

- On a consistent background, a 2px contrasting outline.
- On a consistent background, a 1px contrasting outline with outline padding.
- An internal 3px thick outline.

We should also keep an eye on the [CSS `color-contrast()` function](https://www.w3.org/TR/css-color-5/#colorcontrast), which asks the browser to pick a contrasting color. (Hat tip to the [Shoptalk show](https://shoptalkshow.com/457/#t=00:55).) That might make a good technique if & when it gathers browser support.

**Comment by LiLoDavis** [comment]
One complication I'd like to eliminate: For minimum area, the _outline_ is based on the _perimeter_ of the component, while the _shape_ is based on its _minimum bounding box_. Also, it isn't clear whether the perimeter and bounding box are determined by the portions of the component that are visible, or the portions that are interactive.

**Comment by alastc** [comment]
Just looking at this again, and unfortunately I don't see how to use a simpler size metric without either:
- Disallowing some good indicators, or
- Allowing some very poor indicators.

A change I have made though (in #2000) is to put three good examples at the top of the understanding with the line:

> Any outline that is at least 2px thick and contrasts with the non-focused state would pass this criterion. The detail in this document is for people who do not wish to use such an outline.

**Comment by alastc** [comment]
Proposed response:

<hr>

There are a couple of suggestions here:

<h3>Contrasting area calc</h3>

Reducing the contrasting area to "at least as large as a 3 CSS pixels border along the shortest side of the unfocused component box." would undermine the usefulness of the criterion for the user. 

The bullet itself looks easier, but it becomes harder to apply. The first bullet of the "minimum area" tackles the most common case: an outline. 

Unless someone can come up with a simpler calculation that allows for large+thin indicators and small+thick indicators, it is not an improvement.

In the meantime, we recently updated the [Intent section](https://w3c.github.io/wcag/understanding/focus-appearance-minimum.html#intent) to include some simple examples that meet the criterion.

<h3>Bounding box vs perimeter</h3>

> the outline is based on the perimeter of the component, while the shape is based on its minimum bounding box.

The minimum bounding box is useful for odd shapes, but I wonder if it would make the 1st bullet harder to understand if it were used there? E.g:

"Outline: the area of a 1 CSS pixel thick bounding box of the unfocused component"

<h3>Visible or interactive boundary</h3>

> it isn't clear whether the perimeter and bounding box are determined by the portions of the component that are visible, or the portions that are interactive

Generally it should be whatever is part of that component (i.e. interactive). If it doesn't have a visible boundary (e.g. text-link), the natural outline will be larger than the visible size, so it doesn't clash (i.e. the default outline will be larger than one around the visible area). There is also a note underneath saying:
"If the component has a visible boundary smaller than the hit area, or the size of the component is not available, the minimum area can be taken from the visible boundary."
